### PR TITLE
[flang][MLIR][OpenMP] Extend delayed privatization for scalar allocatables and pointers

### PR DESCRIPTION
**Pull request by ergawy**
One more step in extending support for delayed privatization. This diff adds support for scalar allocatables and pointers.

### CODE REVIEW

**Comment by ergawy** [comment]
This PR replaces the previous overly complicated solution proposed in https://github.com/llvm/llvm-project/pull/84033.

**Comment by kiranchandramohan** [comment]
Thanks @ergawy for the change. And sorry for the changes you had to make due to the previous usage of both base and original base.

I was expecting the deallocation of the allocatable variable to be inside a separate region in the PrivateClause Op.

```
init {
} copy {
} dealloc {
// insert dealloc here
}
```

**Comment by ergawy** [comment]
> ... And sorry for the changes you had to make due to the previous usage of both base and original base.

No problem at all. I learned quite a bit trying to figure out the solution so the time spent did not go to waste for sure.

> I was expecting the deallocation of the allocatable variable to be inside a separate region in the PrivateClause Op.
> 
> ```
> init {
> } copy {
> } dealloc {
> // insert dealloc here
> }
> ```

That should be the case indeed. I added a `TODO` in `DataSharingProcessor::insertDeallocs()` to extend `omp.private` op to add a `dealloc` region since we don't have that one (I did not anticipate the need for it since I was not familiar with how allocatables work when I started). I will create follow up PR(s) for this if you don't mind.

File: flang/test/Lower/OpenMP/delayed-privatization-allocatable-firstprivate.f90 [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+!$omp parallel firstprivate(var1)
+  var1 = 10
+!$omp end parallel
+end subroutine
+
+! CHECK-LABEL: omp.private {type = firstprivate}
+! CHECK-SAME: @[[PRIVATIZER_SYM:.*]] : [[TYPE:!fir.ref<!fir.box<!fir.heap<i32>>>]] alloc {
+
+! CHECK-NEXT: ^bb0(%[[PRIV_ARG:.*]]: [[TYPE]]):
+
```
Comment (kiranchandramohan) [inline flang/test/Lower/OpenMP/delayed-privatization-allocatable-firstprivate.f90:18]: What is allocated is not visible in the test here.

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 2]
```diff
@@ -50,6 +50,7 @@ void DataSharingProcessor::processStep2(mlir::Operation *op, bool isLoop) {
 }
 
 void DataSharingProcessor::insertDeallocs() {
+  // TODO Extend delayed privatization to include a `dealloc` region?
```
Comment (kiranchandramohan) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:53]: Nit: you can remove the question mark.

**Review by kiranchandramohan — COMMENTED** [review]
> I will create follow up PR(s) for this if you don't mind.

That is fine.

File: flang/test/Lower/OpenMP/delayed-privatization-allocatable-firstprivate.f90 (hunk 1 above)
Comment (kiranchandramohan) [reply flang/test/Lower/OpenMP/delayed-privatization-allocatable-firstprivate.f90:18]: Is that deferred to the private test?

**Comment by kiranchandramohan** [comment]
Would pointers come in a separate patch?

File: flang/include/flang/Optimizer/Builder/BoxValue.h [hunk 3]
```diff
@@ -535,6 +535,17 @@ class ExtendedValue : public details::matcher<ExtendedValue> {
...
+  /// Clone an ExtendedValue to a new instance changing the base address.
+  ///
+  /// TODO So far, this is only a shallow clone; only the base address is
+  /// replaced. This will probably be extended to implement deep cloning to
+  /// support scenarios such as delayed privatization for ArrayBoxValue's.
+  ExtendedValue clone(mlir::Value newBase) const {
+    return match(
+        [&](const UnboxedValue &box) -> ExtendedValue { return newBase; },
+        [&](const auto &box) -> ExtendedValue { return box.clone(newBase); });
+  }
```
Comment (kiranchandramohan) [inline flang/include/flang/Optimizer/Builder/BoxValue.h:547]: Can substBase be used here? `lib/Optimizer/Builder/BoxValue.cpp:fir::ExtendedValue fir::substBase`

File: flang/lib/Lower/Bridge.cpp [hunk 4]
```diff
@@ -1060,15 +1058,16 @@ class FirConverter : public Fortran::lower::AbstractConverter {
...
     if (lowerToHighLevelFIR())
-      copyVarHLFIR(loc, lhs_sb.getAddr(), rhs_sb.getAddr());
+      copyVarHLFIR(loc, lhs_sb, rhs_sb);
     else
       copyVarFIR(loc, sym, lhs_sb, rhs_sb);
   }
 
-  void copyVarHLFIR(mlir::Location loc, mlir::Value dst, mlir::Value src) {
+  void copyVarHLFIR(mlir::Location loc, Fortran::lower::SymbolBox dst,
+                    Fortran::lower::SymbolBox src) {
```
Comment (jeanPerier) [inline flang/lib/Lower/Bridge.cpp:1067]: Why did you need to change the argument here and the way allocatable were tested for?

File: flang/test/Lower/OpenMP/delayed-privatization-allocatable-firstprivate.f90 (hunk 1 above)
Comment (ergawy) [reply flang/test/Lower/OpenMP/delayed-privatization-allocatable-firstprivate.f90:18]: Indeed, instead of just copying the check lines from the private test again, I skipped most of it here. Let me know if you prefer to repeat the checks here as well.

File: flang/lib/Lower/Bridge.cpp (hunk 4 above)
Comment (ergawy) [reply flang/lib/Lower/Bridge.cpp:1067]: However, I just realized that for eager privatization, we also have to check for the `FortranVariableOpInterface` of the `SymbolBox` variant. I didn't catch the issue since all tests were green so I will add more tests for eager privatization for both allocatables and pointers in a follow-up PR. Seems like this was an oversight, I think.

**Comment by ergawy** [comment]
kjk

> Would pointers come in a separate patch?

Done in latest commit.

**Comment by ergawy** [comment]
I think this is now ready for another round of review :). I am working on supporting delayed privatization for array at the moment 👨‍🏭 .

**Review by kiranchandramohan — APPROVED** [review]
LG.
